octavia-cli/octavia_cli/apply/resources.py

```diff
@@ -0,0 +1,426 @@
+        )
+
+    @property
+    def remote_resource(self):
+        return self._get_remote_resource()
+
+    @property
+    def was_created(self):
+        return True if self.remote_resource else False
+
+    def __getattr__(self, name: str) -> Any:
+        """Map attribute of the YAML config to the Resource object.
+
+        Args:
+            name (str): Attribute name
+
+        Raises:
+            AttributeError: Raised if the attributed was not found in the local configuration.
+
+        Returns:
+            [Any]: Attribute value
+        """
+        if name in self.local_configuration:
+            return self.local_configuration.get(name)
+        raise AttributeError(f"{self.__class__.__name__}.{name} is invalid.")
+
+    def _search(self) -> Union[SourceReadList, DestinationReadList]:
+        """Run search of a resources on the remote Airbyte instance.
+
+        Returns:
+            Union[SourceReadList, DestinationReadList]: Search results
+        """
+        return self._search_fn(self.api_instance, self.search_payload)
+
+    def _get_state_from_file(self) -> Optional[ResourceState]:
+        """Retrieve a state object from a local YAML file if it exists.
+
+        Returns:
+            Optional[ResourceState]: the deserialized resource state if YAML file found.
+        """
+        expected_state_path = Path(os.path.join(os.path.dirname(self.configuration_path), "state.yaml"))
+        if expected_state_path.is_file():
+            return ResourceState.from_file(expected_state_path)
+
+    def _get_remote_resource(self) -> Optional[Union[SourceRead, DestinationRead]]:
+        """Find the remote resource on the Airbyte instance associated with the current resource.
+
+        Raises:
+            DuplicateRessourceError: raised if the search results return multiple resources.
+
+        Returns:
+            Optional[Union[SourceRead, DestinationRead]]: The remote resource found.
+        """
+        search_results = self._search().get(f"{self.resource_type}s", [])
+        if len(search_results) > 1:
+            raise DuplicateRessourceError("Two or more ressources exist with the same name.")
+        if len(search_results) == 1:
+            return search_results[0]
+        else:
+            return None
+
+    def get_diff_with_remote_resource(self) -> str:
+        """Compute the diff between current resource and the remote resource.
+
+        Raises:
+            NonExistingRessourceError: Raised if the remote resource does not exist.
+
+        Returns:
+            str: The prettyfied diff.
+        """
+        if not self.was_created:
+            raise NonExistingRessourceError("Cannot compute diff with a non existing remote resource.")
+        current_config = self.configuration
+        remote_config = self.remote_resource.connection_configuration
+        diff = compute_diff(remote_config, current_config)
+        return diff.pretty()
+
+    def _create_or_update(
```

Comment:
I think this will all work as expected if the remote Airbyte instance does not have any sources/destinations already that the user wants to interact with with this CLI.

However, what about the case where someone wants to use this CLI to manage sources/destinations that already exist in their Airbyte instance? Is there some way for them to export their existing sources/destinations to create these resources, so that they don't need to recreate them with this CLI? Maybe that functionality is coming in a later PR?

Another way I could see a user getting into this case is if they delete all of the octavia-cli resource files - they should have some way to regenerate them from their Airbyte instance.

Reply:
You're 💯  right! This functionality will come with the `octavia import` feature. It's only possible to update resources created with the CLI at the moment.